fontique/src/backend/coretext.rs

```diff
@@ -77,97 +71,39 @@ impl SystemFonts {
+        let family_name = unsafe { CTFontCopyFamilyName(&font) };
+        self.name_map.get(&family_name.to_string()).map(|n| n.id())
     }
-}
 
-impl SystemFonts {
-    fn fallback_for_text(
+    fn fallback_font_for_text(
         &mut self,
         text: &str,
         locale: Option<&str>,
         prefer_ui: bool,
-    ) -> Option<FamilyId> {
-        let font = fallback_for_text(text, locale, prefer_ui)?;
-        self.name_map.get(&font.family_name()).map(|n| n.id())
-    }
-}
-
-fn fallback_for_text(text: &str, locale: Option<&str>, prefer_ui: bool) -> Option<CTFont> {
-    let cf_text = CFString::new(text);
-    let cf_text_range = CFRange::init(0, cf_text.char_len());
-    let cf_locale = locale.map(CFString::new);
-    let base_font = {
-        let desc_attrs = CFDictionary::from_CFType_pairs(&[]);
-        let mut desc = font_descriptor::new_from_attributes(&desc_attrs);
-        if prefer_ui {
-            if let Some(ui_desc) = ui_font_for_language(kCTFontSystemFontType, 0.0, None)
-                .map(|font| font.copy_descriptor())
-            {
-                desc = ui_desc;
-            }
-        }
-        font::new_from_descriptor(&desc, 0.0)
-    };
-    let font = unsafe {
-        CTFont::wrap_under_create_rule(if let Some(locale) = cf_locale {
-            CTFontCreateForStringWithLanguage(
-                base_font.as_concrete_TypeRef(),
-                cf_text.as_concrete_TypeRef(),
-                cf_text_range,
-                locale.as_concrete_TypeRef(),
-            )
-        } else {
-            CTFontCreateForString(
-                base_font.as_concrete_TypeRef(),
-                cf_text.as_concrete_TypeRef(),
-                cf_text_range,
-            )
-        })
-    };
-    Some(font)
-}
-
-fn ui_font_for_language(
-    ui_type: CTFontUIFontType,
-    size: f64,
-    language: Option<CFString>,
-) -> Option<CTFont> {
-    unsafe {
-        let font_ref = CTFontCreateUIFontForLanguage(
-            ui_type,
-            size,
-            language
-                .as_ref()
-                .map(|x| x.as_concrete_TypeRef())
-                .unwrap_or(std::ptr::null()),
-        );
-        if font_ref.is_null() {
-            None
+    ) -> Option<CFRetained<CTFont>> {
+        let text = CFString::from_str(text);
+        let text_range = CFRange {
+            location: 0,
+            length: unsafe { CFStringGetLength(&text) },
+        };
+        let locale = locale.map(CFString::from_str);
+        let base_font = if prefer_ui {
+            unsafe { CTFontCreateUIFontForLanguage(CTFontUIFontType::System, 0.0, None).unwrap() }
         } else {
-            Some(CTFont::wrap_under_create_rule(font_ref))
-        }
+            unsafe {
+                let attrs = CFDictionaryCreate(None, null_mut(), null_mut(), 0, null(), null());
+                let desc = CTFontDescriptorCreateWithAttributes(&attrs.unwrap());
+                CTFontCreateWithFontDescriptor(&desc, 0.0, null())
```

Comment:
It is interesting that we don't pass any attributes here. What is the font that gets returned from this sort of thing?

Reply:
I believe in practice it gets you 12pt San Francisco (or Helvetica back-when). `CTFontCreateWithFontDescriptor` takes an unspecified descriptor as an opportunity to default on every field.